core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -598,47 +561,71 @@ class GroupMetadataManager(val brokerId: Int,
+              if (group.is(Empty) && !group.hasOffsets) {
+                group.transitionTo(Dead)
 
-                try {
-                  // do not need to require acks since even if the tombstone is lost,
-                  // it will be appended again in the next purge cycle
-                  partition.appendMessagesToLeader(new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, tombstones: _*))
-                  offsetsRemoved += tombstones.size
-                }
-                catch {
-                  case t: Throwable =>
-                    error("Failed to mark %d expired offsets for deletion in %s.".format(tombstones.size, appendPartition), t)
-                  // ignore and continue
+                // We avoid writing the tombstone
+                // when the generationId is 0, since this group is only using Kafka for offset storage.
+                if (groupMetadataCache.get(group.groupId) == group && group.generationId > 0) {
+                  // Append the tombstone messages to the partition. It is okay if the replicas don't receive these (say,
+                  // if we crash or leaders move) since the new leaders will still expire the consumers with heartbeat and
+                  // retry removing this group.
+
+                  trace("Marking group %s as deleted.".format(group.groupId))
+
+                  tombstones += new Message(bytes = null, key = GroupMetadataManager.groupMetadataKey(group.groupId),
+                    timestamp = timestamp, magicValue = magicValue)
                 }
-              }
 
-              if (group.is(Empty) && !group.hasOffsets) {
-                group.transitionTo(Dead)
-                evictGroupAndDeleteMetadata(group)
                 info("Group %s generation %s is dead and removed".format(group.groupId, group.generationId))
               }
 
+              Some((group, offsetsPartition, tombstones))
             case None =>
               info("BrokerId %d is no longer a coordinator for the group %s. Proceeding cleanup for other alive groups".format(brokerId, group.groupId))
+              None
           }
+        } else {
+          None
+        }
+      }
+    }
+
+    for ((group, offsetsPartition, tombstones) <- result) {
+      val partitionOpt = replicaManager.getPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+
+      partitionOpt.foreach { partition =>
+        val appendPartition = TopicAndPartition(Topic.GroupMetadataTopicName, offsetsPartition)
+        trace("Marked %d offsets in %s for deletion.".format(tombstones.size, appendPartition))
+
+        try {
+          // do not need to require acks since even if the tombstone is lost,
+          // it will be appended again in the next purge cycle
+          partition.appendMessagesToLeader(new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, tombstones: _*))
+          offsetsRemoved += tombstones.size
```

Comment:
I think one log append will be slightly more effective than two. The problem here only in info logging. It may be to do so `(group, offsetsPartition, tombstones, expiredOffsetsSize)` ?
```
offsetsRemoved += expiredOffsetsSize
```